Allow custom IExternalScopeProvider in LoggerFactory via DI

heathbm opened the pull request:
Implementation of api-approved issue: #50590

heathbm committed:
Allow custom IExternalScopeProvider in LoggerFactory via DI

KalleOlaviNiemitalo commented on src/libraries/Microsoft.Extensions.Logging/src/LoggerFactory.cs:
```diff
@@ -62,8 +62,21 @@ public LoggerFactory(IEnumerable<ILoggerProvider> providers, IOptionsMonitor<Log
+        /// <param name="providers">The providers to use in producing <see cref="ILogger"/> instances.</param>
+        /// <param name="filterOption">The filter option to use.</param>
+        /// <param name="options">The <see cref="LoggerFactoryOptions"/>.</param>
+        /// <param name="scopeProvider">The <see cref="IExternalScopeProvider"/>.</param>
```
Should this document that LoggerFactoryOptions.ActivityTrackingOptions takes effect only if `scopeProvider` is null?

KalleOlaviNiemitalo replied:
I think it's good to keep validating LoggerFactoryOptions.ActivityTrackingOptions even if `scopeProvider` is not null.

heathbm replied:
> Should this document that LoggerFactoryOptions.ActivityTrackingOptions takes effect only if `scopeProvider` is null?

I see your point. However, I also think it should be safe enough to assume someone who is willing to provide a custom IExternalScopeProvider, wants to do so because they want to manage everything, including trace context.

heathbm replied:
> I think it's good to keep validating LoggerFactoryOptions.ActivityTrackingOptions even if `scopeProvider` is not null.

If `scopeProvider` is not null, `LoggerFactoryOptions.ActivityTrackingOptions` will not be used. Therefore, validating something that is unused and will have no effect seems redundant.

KalleOlaviNiemitalo replied:
Do you intend to make the validation conditional, then?

I expect that changing the validation to `if (scopeProvider is null && (_factoryOptions.ActivityTrackingOptions & ActivityTrackingOptionsMask) != 0)` would just make it slower without any benefit to application developers.

Adding `&& scopeProvider is null` to the

heathbm replied:
Sorry, I misread your comment, I thought you were suggesting adding validation. Although it is redundant validation when `scopeProvider` is null, the argument for keeping it is compelling also: invalid options will be caught even when options are not relevant. Also, it could be risky if future changes cause this assumption to no longer be true, the conditional part of the check could be missed. So, I am in favor of keeping the check.

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging/ref/Microsoft.Extensions.Logging.cs:
```diff
@@ -57,7 +57,9 @@ public partial class LoggerFactory : Microsoft.Extensions.Logging.ILoggerFactory
-        public LoggerFactory(System.Collections.Generic.IEnumerable<Microsoft.Extensions.Logging.ILoggerProvider> providers, Microsoft.Extensions.Options.IOptionsMonitor<Microsoft.Extensions.Logging.LoggerFilterOptions> filterOption, Microsoft.Extensions.Options.IOptions<Microsoft.Extensions.Logging.LoggerFactoryOptions>? options = null) { }
+        public LoggerFactory(System.Collections.Generic.IEnumerable<Microsoft.Extensions.Logging.ILoggerProvider> providers, Microsoft.Extensions.Options.IOptionsMonitor<Microsoft.Extensions.Logging.LoggerFilterOptions> filterOption, Microsoft.Extensions.Options.IOptions<Microsoft.Extensions.Logging.LoggerFactoryOptions>? options) { }
+        public LoggerFactory(System.Collections.Generic.IEnumerable<Microsoft.Extensions.Logging.ILoggerProvider> providers, Microsoft.Extensions.Options.IOptionsMonitor<Microsoft.Extensions.Logging.LoggerFilterOptions> filterOption, Microsoft.Extensions.Options.IOptions<Microsoft.Extensions.Logging.LoggerFactoryOptions>? options = null, Microsoft.Extensions.Logging.IExternalScopeProvider? scopeP ...
+
```
Please use `/t:GenerateReferenceAssemblySource` target to update the ref files as explained in https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md

heathbm replied:
Thank you for the link, I have updated the ref files.

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging/tests/Common/LoggingServiceCollectionExtensionsTest.cs:
```diff
@@ -23,6 +25,43 @@ public void AddLogging_WrapsServiceCollection()
+            var loggerBuilder = services.AddLogging(builder =>
+            {
+                callbackCalled = true;
+                Assert.Same(services, builder.Services);
```
This assertion doesn't seem to be adding value and not relevant to injecting scope provider.

heathbm replied:
I have removed the redundant assertion.

maryamariyan reviewed (commented):
Thanks for the PR

heathbm committed:
Remove redundant assertion

heathbm committed:
Update Microsoft.Extensions.Logging.cs ref

heathbm committed:
Restore previous FilterLoggingBuilderExtensions ref order 

Reverting an auto-generated ref change by /t:GenerateReferenceAssemblySource as it does not appear to be relevant to the scope of changes

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging/tests/Common/LoggingServiceCollectionExtensionsTest.cs:
```diff
@@ -23,6 +25,42 @@ public void AddLogging_WrapsServiceCollection()
+        [Fact]
+        public void AddLogging_TestInjectedScopeProvider()
+        {
+            var testScope = "test scope";
+            var services = new ServiceCollection();
+            var externalScopeProvider = new Mock<IExternalScopeProvider>();
+            var callbackCalled = false;
+            externalScopeProvider
+                .Setup(e => e.Push(It.IsAny<object?>()))
+                .Callback(() => callbackCalled = true);
+
+            services.AddSingleton(externalScopeProvider.Object);
+            var loggerBuilder = services.AddLogging();
+            var serviceProvider = services.BuildServiceProvider();
+            var loggerFactory = serviceProvider.GetService<ILoggerFactory>();
+            var logger = loggerFactory.CreateLogger<LoggingServiceCollectionExtensionsTest>();
+            logger.BeginScope(testScope);
+
+            Assert.True(callbackCalled);
```
nit suggestion

```suggestion
            bool callbackCalled = false;
            var externalScopeProvider = new Mock<IExternalScopeProvider>();
            externalScopeProvider
                .Setup(e => e.Push(It.IsAny<object?>()))
                .Callback(() => callbackCalled = true);
      
            var serviceProvider = new ServiceCollection()
                .AddSingleton(externalScopeProvider.Object)
                .AddLogging()
                .BuildServiceProvider();

            var loggerFactory = serviceProvider.GetService<ILoggerFactory>();
            var logger = loggerFactory.CreateLogger<LoggingServiceCollectionExtensionsTest>();
            logger.BeginScope("test scope");

            Assert.True(callbackCalled);
```

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging/tests/Common/LoggingServiceCollectionExtensionsTest.cs:
```diff
@@ -23,6 +25,42 @@ public void AddLogging_WrapsServiceCollection()
+        [Fact]
+        public void AddLogging_TestInjectedScopeProvider()
+        {
+            var testScope = "test scope";
+            var services = new ServiceCollection();
+            var externalScopeProvider = new Mock<IExternalScopeProvider>();
+            var callbackCalled = false;
+            externalScopeProvider
+                .Setup(e => e.Push(It.IsAny<object?>()))
+                .Callback(() => callbackCalled = true);
+
+            services.AddSingleton(externalScopeProvider.Object);
+            var loggerBuilder = services.AddLogging();
+            var serviceProvider = services.BuildServiceProvider();
+            var loggerFactory = serviceProvider.GetService<ILoggerFactory>();
+            var logger = loggerFactory.CreateLogger<LoggingServiceCollectionExtensionsTest>();
+            logger.BeginScope(testScope);
+
+            Assert.True(callbackCalled);
```
nit suggestion

```suggestion
            bool callbackCalled = false;
            var externalScopeProvider = new Mock<IExternalScopeProvider>();
            externalScopeProvider
                .Setup(e => e.Push(It.IsAny<object?>()))
                .Callback(() => callbackCalled = true);
      
            var serviceProvider = new ServiceCollection()
                .AddSingleton(externalScopeProvider.Object)
                .AddLogging()
                .BuildServiceProvider();

            var loggerFactory = serviceProvider.GetService<ILoggerFactory>();
            var logger = loggerFactory.CreateLogger<LoggingServiceCollectionExtensionsTest>();
            logger.BeginScope("test scope");

            Assert.True(callbackCalled);
```

heathbm replied:
Thanks for the suggestion, I've applied it

maryamariyan reviewed (approved):
LGTM with comments

heathbm committed:
Update AddLogging_TestInjectedScopeProvider test format

Co-authored-by: Maryam Ariyan <maryam.ariyan@microsoft.com>

heathbm committed:
Remove redundant test: AddLogging_InjectScopeProvider

heathbm commented on src/libraries/Microsoft.Extensions.Logging/tests/Common/LoggingServiceCollectionExtensionsTest.cs:
```diff
@@ -23,6 +25,42 @@ public void AddLogging_WrapsServiceCollection()
         }
 
+        [Fact]
+        public void AddLogging_InjectScopeProvider()
```
I agree, removed

maryamariyan commented:
Thanks for the contribution @heathbm
